sql: add not visible index to optimizer

wenyihu6 opened the pull request:
This commit adds the logic of the invisible index feature to the optimizer.
After this commit has been merged, the invisible index feature should be fully
functional with `CREATE INDEX` and `CREATE TABLE`.

Tested with TPC-H: marking indexes invisible produced identical query plans as dropping indexes.

<img width="938" alt="Screen Shot 2022-09-09 at 1 24 03 PM" src="https://user-images.githubusercontent.com/56973754/189409225-4560ac96-b7ff-46f6-b606-8718940bbe6e.png">

Assists: https://github.com/cockroachdb/cockroach/issues/72576

See also: https://github.com/cockroachdb/cockroach/pull/85239

Release note (sql change): creating a not visible index using
`CREATE TABLE …(INDEX … NOT VISIBLE)` or
`CREATE INDEX … NOT VISIBLE` is now supported.

wenyihu6 committed:
sql: add not visible index to optimizer

This commit adds the logic of the invisible index feature to the optimizer.
After this commit has been merged, the invisible index feature should be fully
functional with `CREATE INDEX` and `CREATE TABLE`.

Assists: https://github.com/cockroachdb/cockroach/issues/72576

See also: https://github.com/cockroachdb/cockroach/pull/85239

Release note (sql change): creating a not visible index using
`CREATE TABLE …(INDEX … NOT VISIBLE)` or
`CREATE INDEX … NOT VISIBLE` is now supported.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85794)
<!-- Reviewable:end -->

wenyihu6 commented:
I split the other PR https://github.com/cockroachdb/cockroach/pull/85354. This PR contains changes in the optimizer, and the other PR only has notice related code.

mgartner reviewed (commented):
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85794)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @wenyihu6)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85794#-N97ly_w4jE_sRHexHK2:-N97ly_w4jE_sRHexHK3:b2ptz8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/272799e8da60efa980fef81ecb78147a1832a2d3/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L192)):*
> ```
> # After making idx_v_invisible invisible, SELECT ignores idx_v_invisible.
> query T
> EXPLAIN SELECT v FROM t1 WHERE v = 2
> ```

nit: add a test just like this with the index forced - you already have a similar tests for partial indexes below, but I think it'd be nice to have one for non-partial indexes here.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85794#-N97mWSJ9hh9BxfNMWAH:-N97mWSJ9hh9BxfNMWAI:b3s5kti) ([raw file](https://github.com/cockroachdb/cockroach/blob/272799e8da60efa980fef81ecb78147a1832a2d3/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L391)):*
> ```
> #################################################################################
> # These tests check for some other cases called from buildDataSource (buildJoin,
> # buildInputForUpdate, buildFromWithLateral).
> ```

I'm not sure I understand the motivation for testing based on the sources. These functions should end up calling buildScan in the optbuilder phase. A more interesting test might be testing lookup join, since they also iterate over seconary indexes. For example: https://github.com/cockroachdb/cockroach/blob/14a0ef30a8680f2eeeef3a3f59c816dfaf9cf836/pkg/sql/opt/exec/execbuilder/testdata/lookup_join#L44-L60

Generation of zigzag joins and merge joins also iterate over secondary indexes, so those could be interesting tests too.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 793 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85794#-N97pjNS2MSCS4HuoWNN:-N97pjNS2MSCS4HuoWNO:btkctua) ([raw file](https://github.com/cockroachdb/cockroach/blob/272799e8da60efa980fef81ecb78147a1832a2d3/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L793)):*
> ```
> # c_update_idx_invisible is used for constraint check (delete on a parent table requires FK check).
> query T
> EXPLAIN DELETE FROM parent where p = 2
> ```

There's a [delete cascade fast path](https://github.com/cockroachdb/cockroach/blob/8004e0d6e78da30f7d329008c485e6f5abaa6b0b/pkg/sql/opt/optbuilder/fk_cascade.go#L116-L140) that we should test here. Notice the slight difference in the two query plans below. The first's `fk-cascade` does not scan `buffer 1`:

```
defaultdb> create table p (p int primary key, i int);
CREATE TABLE

defaultdb> create table c (c int primary key, p int references
p(p) on delete cascade);
CREATE TABLE

defaultdb> explain delete from p where p = 1;
           info
--------------------------
  distribution: local
  vectorized: true

  • root
  │
  ├── • delete range
  │     from: p
  │     spans: [/1 - /1]
  │
  └── • fk-cascade
        fk: c_p_fkey
(11 rows)

defaultdb> explain delete from p where p = 1 and i = 2;
                   info
------------------------------------------
  distribution: local
  vectorized: true

  • root
  │
  ├── • delete
  │   │ from: p
  │   │
  │   └── • buffer
  │       │ label: buffer 1
  │       │
  │       └── • filter
  │           │ filter: i = 2
  │           │
  │           └── • scan
  │                 missing stats
  │                 table: p@p_pkey
  │                 spans: [/1 - /1]
  │
  └── • fk-cascade
        fk: c_p_fkey
        input: buffer 1
```

I think you'll need to run this delete before adding the `child_update` table to get the delete fast path to trigger.
___
*[`pkg/sql/opt/xform/scan_index_iter.go` line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85794#-N91r_eh07BFzBk5b2Xy:-N91r_eh07BFzBk5b2Xz:b71i15i) ([raw file](https://github.com/cockroachdb/cockroach/blob/272799e8da60efa980fef81ecb78147a1832a2d3/pkg/sql/opt/xform/scan_index_iter.go#L230)):*
> ```Go
> 		} else {
> 			// If we are not forcing any specific index and not visible index feature is
> 			// enabled here, ignore not visible indexes only.
> ```

nit: remove "only"


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85794#-N97ly_w4jE_sRHexHK2:-N99hQ2KDSb_G1LWscdI:bsxshjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/272799e8da60efa980fef81ecb78147a1832a2d3/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L192)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a test just like this with the index forced - you already have a similar tests for partial indexes below, but I think it'd be nice to have one for non-partial indexes here.
</blockquote></details>

I think there is a case like this below (the first case under force index) https://github.com/cockroachdb/cockroach/blob/4a2af4e2aa538cca07e92b9f768141030afd7b0b/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L347


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85794#-N97mWSJ9hh9BxfNMWAH:-N99iFjHE2TMlmAR2SOm:b98yl4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/272799e8da60efa980fef81ecb78147a1832a2d3/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L391)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm not sure I understand the motivation for testing based on the sources. These functions should end up calling buildScan in the optbuilder phase. A more interesting test might be testing lookup join, since they also iterate over seconary indexes. For example: https://github.com/cockroachdb/cockroach/blob/14a0ef30a8680f2eeeef3a3f59c816dfaf9cf836/pkg/sql/opt/exec/execbuilder/testdata/lookup_join#L44-L60

Generation of zigzag joins and merge joins also iterate over secondary indexes, so those could be interesting tests too.


</blockquote></details>

My initial intention was that we have covered most of places where buildScan was called now. buildDataSource was the only function where some cases were not covered like (buildJoin, buildFromWithLateral). I've removed these cases and added more cases to cover join and one case for [lateral subquery](https://github.com/cockroachdb/cockroach/blob/4a2af4e2aa538cca07e92b9f768141030afd7b0b/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L512)


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 793 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85794#-N97pjNS2MSCS4HuoWNN:-N99jCzG-umeRJWp_EW3:b-wsl87l) ([raw file](https://github.com/cockroachdb/cockroach/blob/272799e8da60efa980fef81ecb78147a1832a2d3/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L793)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

There's a [delete cascade fast path](https://github.com/cockroachdb/cockroach/blob/8004e0d6e78da30f7d329008c485e6f5abaa6b0b/pkg/sql/opt/optbuilder/fk_cascade.go#L116-L140) that we should test here. Notice the slight difference in the two query plans below. The first's `fk-cascade` does not scan `buffer 1`:

```
defaultdb> create table p (p int primary key, i int);
CREATE TABLE

defaultdb> create table c (c int primary key, p int references
p(p) on delete cascade);
CREATE TABLE

defaultdb> explain delete from p where p = 1;
           info
--------------------------
  distribution: local
  vectorized: true

  • root
  │
  ├── • delete range
  │     from: p
  │     spans: [/1 - /1]
  │
  └── • fk-cascade
        fk: c_p_fkey
(11 rows)

defaultdb> explain delete from p where p = 1 and i = 2;
                   info
------------------------------------------
  distribution: local
  vectorized: true

  • root
  │
  ├── • delete
  │   │ from: p
  │   │
  │   └── • buffer
  │       │ label: buffer 1
  │       │
  │       └── • filter
  │           │ filter: i = 2
  │           │
  │           └── • scan
  │                 missing stats
  │                 table: p@p_pkey
  │                 spans: [/1 - /1]
  │
  └── • fk-cascade
        fk: c_p_fkey
        input: buffer 1
```

I think you'll need to run this delete before adding the `child_update` table to get the delete fast path to trigger.
</blockquote></details>

I think this case is triggering fast path because the filter on p got transferred to the cascade constraint. I added another case to cover when fast path is not triggered.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> :100: Nice work, Wenyi!

> Reviewed 2 of 4 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85794#-:-N9AbCy51iKCODblNzzG:b-3eqkqf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTRs! bors r+

wenyihu6 commented:
retrying ...

bors r+
